File: ruby/plus-one.rb
#!/usr/bin/env ruby

# Given a non-negative integer represented as a non-empty array of digits, plus one to the integer.

# You may assume the integer do not contain any leading zero, except the number 0 itself.

# The digits are stored such that the most significant digit is at the head of the list.

# @param {Integer[]} digits
# @return {Integer[]}
def plus_one(digits)
  pre = 1
  digits.to_enum.with_index.reverse_each do |digit, idx|
    break if pre == 0
    pre, res = (digit + pre).divmod 10
    digits[idx] = res
  end

  digits.unshift(pre) unless pre == 0
    
  digits
end

digits = ARGV[0].chars.map{|c| c.to_i}
puts "digits are #{digits}"
puts "res is #{plus_one(digits)}"

File: swift/t_code/Source/Easy/PlusOne.swift
//
//  PlusOne.swift
//  t_code
//
//  Created by tcui on 9/5/2018.
//  Copyright © 2018 LuckyTR. All rights reserved.
//

import Foundation

/*
     Given a non-empty array of digits representing a non-negative integer, plus one to the integer.
 
     The digits are stored such that the most significant digit is at the head of the list, and each element in the array contain a single digit.
 
     You may assume the integer does not contain any leading zero, except the number 0 itself.
 
     Example 1:
 
     Input: [1,2,3]
     Output: [1,2,4]
     Explanation: The array represents the integer 123.
     Example 2:
 
     Input: [4,3,2,1]
     Output: [4,3,2,2]
     Explanation: The array represents the integer 4321.
 
 */

class PlusOne {
    class Solution {
        func plusOne(_ digits: [Int]) -> [Int] {
            var carry = false
            
            var index = digits.count - 1
            var res = digits
            
            while index >= 0 {
                let add = digits[index] + 1
                if add < 10 {
                    res[index] = add
                    return res
                } else {
                    res[index] = 0
                    carry = true
                }
                index -= 1
            }
            
            if carry {
                res.insert(1, at: 0)
            }
            
            return res
        }
    }
}
